### DISCUSSION TITLE
Flash for variable storage on RP2040

### DISCUSSION THREAD

**Discussion by jdh1955** [tabemann/zeptoforth#128] [Q&A]
On an RPi Pico I would like to use flash to store some calibration variables. I found the block module, can save and list blocks but I don't want to evaluate them with load, I would just like to use block! to save 1024 bytes and recover them as bytes which I will use for variable storage. Perhaps there is another way to access the flash. I see there is a block@ in the block-dev module but I can't see how to point to the on board flash. Is there a value for dev to identify the on board flash? Any help would be gratefully received.
jdh1955

**Comment by tabemann** [comment]
On Mon, Aug 5, 2024 at 9:28 AM jdh1955 ***@***.***> wrote:

> On an RPi Pico I would like to use flash to store some calibration
> variables. I found the block module, can save and list blocks but I don't
> want to evaluate them with load, I would just like to use block! to save
> 1024 bytes and recover them as bytes which I will use for variable storage.
> Perhaps there is another way to access the flash. I see there is a block@
> in the block-dev module but I can't see how to point to the on board flash.
> Is there a value for dev to identify the on board flash? Any help would be
> gratefully received.
> jdh1955
>
The simplest way of doing this (note that this is not compatible with
having FAT32 filesystems in block storage) is using BLOCK::FIND-BLOCK ( id
-- addr | 0 ) to find an address of a 1024 byte buffer containing the
contents of a block with the specified id (or 0 if the block does not
exist) and BLOCK::BLOCK! ( addr id -- ) to write 1024 bytes from the buffer
pointed to by the provided address to the block with the specified id (it
does not need to previously exist). One note is that copying the data from
the buffer returned by BLOCK::FIND-BLOCK directly into another block with
BLOCK::BLOCK! will result in undefined behavior. Also note that
BLOCK::BLOCK! is not realtime-friendly, so you should not use it in any
realtime code. Also note that all id's are valid other than $FFFFFFFF, and
block storage properly acts as a map of id's to 1024 byte blocks rather
than as an array of blocks.

Travis

**Comment by jdh1955** [comment]
Many thanks for the quick response. This is very helpful and does what I want. I’ll also take heed of your warnings.
Regards Jonathan
